src/SimplePDPServiceWithPayments.sol

```diff
@@ -0,0 +1,778 @@
+        uint256 ratePerEpoch = sizeInMiB * RATE_PER_MIB_PER_EPOCH;
+
+        return ratePerEpoch;
+    }
+
+    /**
+     * @notice Decode extra data for proof set creation
+     * @param extraData The encoded extra data from PDPVerifier
+     * @return decoded The decoded ProofSetCreateData struct
+     */
+    function decodeProofSetCreateData(
+        bytes calldata extraData
+    ) internal pure returns (ProofSetCreateData memory) {
+        return abi.decode(extraData, (ProofSetCreateData));
+    }
+
+    /**
+     * @notice Get the total size of a proof set in bytes
+     * @param leafCount Number of leaves in the proof set
+     * @return totalBytes Total size in bytes
+     */
+    function getProofSetSizeInBytes(
+        uint256 leafCount
+    ) internal pure returns (uint256) {
+        return leafCount * BYTES_PER_LEAF;
+    }
+
+    // --- Public getter functions ---
+
+    /**
+     * @notice Get the payment rail ID for a proof set
+     * @param proofSetId The ID of the proof set
+     * @return The payment rail ID, or 0 if not found
+     */
+    function getProofSetRailId(
+        uint256 proofSetId
+    ) external view returns (uint256) {
+        return proofSetInfo[proofSetId].railId;
+    }
+
+    /**
+     * @notice Get payer and payee addresses for a proof set
+     * @param proofSetId The ID of the proof set
+     * @return payer The address paying for storage
+     * @return payee The address receiving payments (SP beneficiary)
+     */
+    function getProofSetParties(
+        uint256 proofSetId
+    ) external view returns (address payer, address payee) {
+        ProofSetInfo storage info = proofSetInfo[proofSetId];
+        return (info.payer, info.payee);
+    }
+
+    /**
+     * @notice Get the metadata for a proof set
+     * @param proofSetId The ID of the proof set
+     * @return The metadata string
+     */
+    function getProofSetMetadata(
+        uint256 proofSetId
+    ) external view returns (string memory) {
+        return proofSetInfo[proofSetId].metadata;
+    }
+
+    /**
+     * @notice Get the metadata for a specific root
+     * @param proofSetId The ID of the proof set
+     * @param rootId The ID of the root
+     * @return The metadata string for the root
+     */
+    function getRootMetadata(
+        uint256 proofSetId,
+        uint256 rootId
+    ) external view returns (string memory) {
+        return proofSetInfo[proofSetId].rootIdToMetadata[rootId];
+    }
+
+    /**
+     * @notice Arbitrates payment based on faults in the given epoch range
+     * @dev Implements the IArbiter interface function
+     * @param railId ID of the payment rail
+     * @param proposedAmount The originally proposed payment amount
+     * @param fromEpoch Starting epoch (exclusive)
+     * @param toEpoch Ending epoch (inclusive)
+     * @return result The arbitration result with modified amount and settlement information
+     */
+    function arbitratePayment(
```

Comment:
@ZenGround0 Can you deeply review this function to ensure it is correct ? Lots of potential for off by ones/edge cases here. Please.